cli/cli.go

```diff
@@ -0,0 +1,168 @@
+		err = unmarshalRequest(ctx, &request)
+		if err == nil {
+			logger.Debugf("executing %s plugin's GetMetadata function", reflect.TypeOf(c.pl))
+			resp, err = c.pl.GetMetadata(ctx, &request)
+		}
+	case plugin.CommandGenerateEnvelope:
+		var request plugin.GenerateEnvelopeRequest
+		err = unmarshalRequest(ctx, &request)
+		if err == nil {
+			logger.Debugf("executing %s plugin's GenerateEnvelope function", reflect.TypeOf(c.pl))
+			resp, err = c.pl.GenerateEnvelope(ctx, &request)
+		}
+	case plugin.CommandVerifySignature:
+		var request plugin.VerifySignatureRequest
+		err = unmarshalRequest(ctx, &request)
+		if err == nil {
+			logger.Debugf("executing %s plugin's VerifySignature function", reflect.TypeOf(c.pl))
+			resp, err = c.pl.VerifySignature(ctx, &request)
+		}
+	case plugin.CommandDescribeKey:
+		var request plugin.DescribeKeyRequest
+		err = unmarshalRequest(ctx, &request)
+		if err == nil {
+			logger.Debugf("executing %s plugin's DescribeKey function", reflect.TypeOf(c.pl))
+			resp, err = c.pl.DescribeKey(ctx, &request)
+		}
+	case plugin.CommandGenerateSignature:
+		var request plugin.VerifySignatureRequest
+		err = unmarshalRequest(ctx, &request)
+		if err == nil {
+			logger.Debugf("executing %s plugin's VerifySignature function", reflect.TypeOf(c.pl))
+			resp, err = c.pl.VerifySignature(ctx, &request)
+		}
+	case plugin.Version:
+		rescueStdOut()
+		c.printVersion(ctx)
+	default:
+		// should never happen
+		rescueStdOut()
+		deliverError(plugin.NewGenericError("something went wrong").Error())
+	}
+
+	op, pluginErr := marshalResponse(ctx, resp, err)
+	rescueStdOut()
+	if pluginErr != nil {
+		deliverError(pluginErr.Error())
+	}
+	fmt.Println(op)
+}
+
+// printHelp prints help text for executable
+func (c Cli) printHelp(ctx context.Context) {
+	md := getMetadata(ctx, c.pl)
+	args := getValidArgsString(md)
+
+	fmt.Printf("%s - %s\n Usage:%s %s", md.Name, md.Description, c.syntax, args)
+}
+
+// printVersion prints version of executable
+func (c Cli) printVersion(ctx context.Context) {
+	md := getMetadata(ctx, c.pl)
+
+	fmt.Printf("%s - %s\nVersion: %s", md.Name, md.Description, md.Version)
+	os.Exit(0)
+}
+
+// validateArgs validate commands/arguments passed to executable.
+func (c Cli) validateArgs(ctx context.Context, args []string) {
+	md := getMetadata(ctx, c.pl)
+	if !(len(args) == 2 && slices.Contains(getValidArgs(md), args[1])) {
+		deliverError(fmt.Sprintf("Invalid command, valid choices are %s %s", c.syntax, getValidArgsString(md)))
+	}
+}
+
+// deferStdout is used to make sure that nothing get emitted to stdout and stderr until intentionally rescued.
+// This is required to make sure that the plugin or its dependency doesn't interferes with notation <-> plugin communication
+func deferStdout() func() {
+	null, _ := os.Open(os.DevNull)
+	sout := os.Stdout
+	serr := os.Stderr
+	os.Stdout = null
+	os.Stderr = null
+
+	return func() {
+		err := null.Close()
+		if err != nil {
+			return
```

Comment:
Should it panic instead of return?

Reply:
Nop, see https://github.com/notaryproject/notation-plugin-framework-go/pull/2#discussion_r1297939515 for reason.

Reply:
When `null, _ := os.Open(os.DevNull)` fails, `null` is `nil`. In this case, calling `null.Close()` will result in `panic()` on calling method on nil pointer.

Reply:
As per [`file_posix.go#Close()`](https://github.com/golang/go/blob/master/src/os/file_posix.go#L19-L24) and I was able to repro the same result in [playground](https://go.dev/play/p/ybdJIMtuR4r)